File: rust/tonk-prose/src-js/document.ts
// Document mode: the element's half of the automerge document protocol.
//
// The element holds NO automerge. The host owns the document; the
// element owns a piece of content and the HEADS it last saw, and
// exchanges both:
//
//   read  -> { heads, content }                 the branch's version
//   write -> send { heads: known, content }     "my content, made on
//                                               top of these heads"
//         <- { local, heads, content }          `local` = the heads my
//                                               content now IS;
//                                               `heads`/`content` = the
//                                               branch after merging
//
// The host applies a write ON TOP OF the heads it names, so a write
// made from a stale view merges instead of overwriting. Two rules keep
// the editor and the branch from fighting:
//
//   1. Remote content is only ever applied when the editor's content is
//      exactly what the host last accounted for. If the user typed
//      during a round trip, the reply's content is dropped and only
//      `local` is kept: the next write then starts from the heads that
//      the previously sent content corresponds to, so nothing is applied
//      twice and nothing is lost.
//   2. One write is in flight at a time. Edits made meanwhile ride the
//      next one.
//
// Pure logic: the transport and the editor are injected, so the node
// tests cover every interleaving. The same file is copied into
// tonk-table, the way `hlc.ts` and `content.ts` are.

/** What the host answers. `T` is the content: markdown for a text
 *  document, a workbook for a table. */
export interface Reply<T> {
  /** The heads the branch is at. */
  heads: string[];
  /** After a write: the heads the content I sent corresponds to. */
  local?: string[];
  /** The content at `heads`. */
  content: T;
  /** The format rule: the document is in a format newer than this app
   *  knows. It can be shown and must not be edited. */
  readonly?: boolean;
  /** After a write: the stored document passed the size at which the
   *  host asks for a warning. Past the hard limit a write is refused. */
  large?: boolean;
}

/** How the session reaches the host. */
export interface Transport<T, E> {
  read(): Promise<Reply<T>>;
  write(heads: string[], edits: E[], request?: EditRequest): Promise<Reply<T>>;
}

export interface EditRequest { id: string; time: number }
function requestIdentity(): EditRequest {
  return { id: crypto.randomUUID(), time: Math.floor(Date.now() / 1000) };
}

/** What the session needs from its editor. */
export interface Editor<T, E> {
  /** The editor's content right now. */
  current(): T;
  /** Replace the editor's content with remote content. */
  apply(content: T): void;
  /** Whether two contents are the same. */
  same(a: T, b: T): boolean;
  /** The edits that turn `from` into `to`. */
  edits(from: T, to: T): E[];
}

export class DocumentSession<T, E> {
  readonly #transport: Transport<T, E>;
  readonly #editor: Editor<T, E>;

  /** The heads the content in `#accounted` corresponds to. */
  #known: string[] = [];
  /** The content the host last accounted for: what a write diffs
   *  against, and the only state remote content may replace. */
  #accounted: T | null = null;
  #inFlight = false;
  #flight: Promise<void> | null = null;
  #pending: { heads: string[]; edits: E[]; content: T; request: EditRequest } | null = null;
  #final: { content: T } | null = null;
  #again = false;
  #closed = false;
  /** Pinned to a past version: opened once, then neither written nor
   *  polled. The transport reads that version; the session only makes
   *  sure nothing the editor holds can leave. */
  readonly #pinned: boolean;
  readonly #onOpen: (() => void) | undefined;
  readonly #onLarge: (() => void) | undefined;
  #large = false;
  /** The host said the document is in a newer format. */
  #newer = false;

  constructor(
    transport: Transport<T, E>,
    editor: Editor<T, E>,
    options: { pinned?: boolean; onOpen?: () => void; onLarge?: () => void } = {},
  ) {
    this.#transport = transport;
    this.#editor = editor;
    this.#pinned = options.pinned === true;
    this.#onOpen = options.onOpen;
    this.#onLarge = options.onLarge;
  }

  get pinned(): boolean {
    return this.#pinned;
  }

  /** Whether the document takes no edits from this element: a past
   *  version, or a format newer than this app knows. */
  get readonly(): boolean {
    return this.#pinned || this.#newer;
  }

  /** The heads this element is at. */
  get heads(): readonly string[] {
    return this.#known;
  }

  get opened(): boolean {
    return this.#accounted !== null;
  }

  /** Load the branch's version into the editor. */
  async open(): Promise<void> {
    const reply = await this.#transport.read();
    if (this.#closed || this.opened) return;
    this.#known = reply.heads;
    this.#accounted = reply.content;
    this.#newer = reply.readonly === true;
    this.#editor.apply(reply.content);
    this.#onOpen?.();
  }

  /** Whether the editor holds edits the host has not accounted for. */
  get dirty(): boolean {
    return (
      this.#pending !== null || (this.#accounted !== null &&
      !this.#editor.same(this.#current(), this.#accounted))
    );
  }

  /** Send the editor's unsent edits. Safe to call at any time and any
   *  number of times; concurrent calls coalesce. */
  flush(): Promise<void> {
    if (this.#flight) { this.#again = true; return this.#flight; }
    this.#flight = this.#flush().finally(() => { this.#flight = null; });
    return this.#flight;
  }

  #current(): T { return this.#final ? this.#final.content : this.#editor.current(); }

  async #flush(): Promise<void> {
    if (this.#closed || this.readonly || this.#accounted === null) return;
    if (this.#inFlight) {
      this.#again = true;
      return;
    }
    const current = this.#current();
    if (!this.#pending && this.#editor.same(current, this.#accounted)) return;
    const pending = this.#pending ??= { heads: [...this.#known], content: current,
      edits: this.#editor.edits(this.#accounted, current), request: requestIdentity() };
    const sent = pending.content;

    this.#inFlight = true;
    try {
      const reply = await this.#transport.write(pending.heads, pending.edits, pending.request);
      if (this.#closed) return;
      this.#pending = null;
      if (reply.large === true && !this.#large) {
        // Said once: automerge keeps every edit, so it only gets larger.
        this.#large = true;
        this.#onLarge?.();
      }
      if (this.#editor.same(this.#current(), sent)) {
        // Nothing typed meanwhile: take the merged branch state.
        this.#known = reply.heads;
        this.#accounted = reply.content;
        if (this.#final) this.#final.content = reply.content;
        if (!this.#final && !this.#editor.same(reply.content, sent)) {
          this.#editor.apply(reply.content);
        }
      } else {
        // Typed during the round trip: keep the editor as it is. The
        // content we sent now IS `local`, so the next write diffs from
        // it and starts at those heads.
        this.#known = reply.local ?? reply.heads;
        this.#accounted = sent;
        this.#again = true;
      }
    } finally {
      this.#inFlight = false;
    }
    if (this.#again) {
      this.#again = false;
      await this.#flush();
    }
  }

  /** Look for changes made elsewhere. Skipped while the editor holds
   *  unsent edits — those go first, and their reply brings the rest. */
  async poll(): Promise<void> {
    if (this.#closed) return;
    // The first read failed — the heads can arrive before their bytes,
    // and the host may not be up yet. Try again; until it works the
    // element shows nothing and takes no edits.
    if (this.#accounted === null) return this.open();
    if (this.#pinned) return;
    if (this.#inFlight || this.dirty) return;
    const known = this.#known;
    const reply = await this.#transport.read();
    if (this.#closed || this.#inFlight || this.dirty || this.#known !== known) return;
    if (sameHeads(reply.heads, this.#known)) return;
    this.#known = reply.heads;
    this.#accounted = reply.content;
    this.#editor.apply(reply.content);
  }

  close(): void {
    this.#closed = true;
  }

  /** Snapshot final text before the editor is destroyed; drain a write
   * already in flight AND the keystrokes made since that write. */
  async finish(): Promise<void> {
    this.#final = { content: this.#editor.current() };
    await this.flush();
    this.close();
  }
}

/** The heads an `at` attribute names: change hashes separated by spaces. */
export function parseHeads(text: string | null): string[] {
  return (text ?? "").split(/\s+/).filter((head) => head !== "");
}

export function sameHeads(a: readonly string[], b: readonly string[]): boolean {
  if (a.length !== b.length) return false;
  const left = [...a].sort();
  const right = [...b].sort();
  return left.every((head, index) => head === right[index]);
}

/** A wire edit of a text document. */
export type TextEdit = { edit: "set-text"; text: string };

/** The editor half for a text document. */
export function textEditor(
  current: () => string,
  apply: (text: string) => void,
): Editor<string, TextEdit> {
  return {
    current,
    apply,
    same: (a, b) => a === b,
    // The whole text: the host diffs it against the text at the heads
    // the write names, which is what the element last accounted for.
    edits: (_from, to) => [{ edit: "set-text", text: to }],
  };
}

/** Reach the host through a bubbling `tonk-document` event, the way
 *  every tonk element reaches it: the host (or, in a sealed guest, its
 *  relay) resolves the repository and branch from the element's routing
 *  context and answers on `detail.result`. */
export function eventTransport<T>(
  element: HTMLElement,
  entity: string,
  format: string,
  content: (body: Record<string, unknown>) => T,
  at: string[] = [],
): Transport<T, unknown> {
  let route: Record<string, unknown> = {};
  const call = async (write?: { heads: string[]; edits: unknown[]; request?: EditRequest }): Promise<Reply<T>> => {
    const detail: Record<string, unknown> = { ...(element.isConnected ? {} : route), entity, format };
    if (write) detail.write = { ...write, format };
    // A past version: the host reads at these heads instead of the branch's.
    else if (at.length > 0) detail.heads = at;
    const event = new CustomEvent("tonk-document", {
      detail,
      bubbles: true,
      composed: true,
      cancelable: true,
    });
    (element.isConnected ? element : element.ownerDocument).dispatchEvent(event);
    if (!event.defaultPrevented || !(detail.result instanceof Promise)) {
      throw new Error("tonk-document: no host answered");
    }
    route = { space: detail.space, branch: detail.branch, profile: detail.profile };
    const body = (await detail.result) as Record<string, unknown>;
    return {
      heads: (body.heads as string[]) ?? [],
      local: body.local as string[] | undefined,
      content: content(body),
      readonly: body.readonly === true,
      large: body.large === true,
    };
  };
  return {
    read: () => call(),
    write: (heads, edits, request) => call({ heads, edits, request }),
  };
}

File: rust/tonk-worker/src/router/document.rs
//! Automerge documents: the element route, the command providers and
//! the sync hook.
//!
//! All behaviour lives in the host-neutral `tonk-document` crate; this
//! module only decides WHEN the worker calls it. The CLI calls the same
//! functions from its own write path, so a document command means the
//! same thing in both.
//!
//! The route is data plane, the class the pinned route table allows: an
//! element moves its content and the heads it last saw, with no user
//! intent to record as a fact. It is fixed to the document cell of one
//! entity, so a page can never reach a branch pointer through it, and
//! its path sits under `/api/repository/{repo}/branch/{branch}/…`, which
//! the portal fetch relay already gates by the guest's own reach.
//!
//! Elements hold no automerge. An element sends its content with the
//! heads it last saw; the edit is applied ON TOP OF THOSE HEADS, merged
//! with whatever the branch gained meanwhile, and the reply carries the
//! merged content plus `local` — the heads the element's own content now
//! corresponds to. An element that kept typing during the round trip
//! sends again from `local`, so nothing is applied twice and nothing is
//! lost.

use std::collections::{HashMap, HashSet};
use std::sync::{Arc, Mutex, OnceLock};

use ::axum::{
    Json,
    body::Bytes,
    extract::{Path, Query, State},
    http::{HeaderMap, StatusCode, header},
    response::{IntoResponse, Response},
};
use axum_wasm_macros::wasm_compat;
use dialog_artifacts::Entity;
use serde::{Deserialize, Serialize};
#[cfg(all(target_arch = "wasm32", target_os = "unknown"))]
use tokio::sync::oneshot;
use tonk_common::log;
use tonk_document::engine::{Content, Edit, Format, Stamp, Table};
use tonk_document::session::{self, Snapshot};

use super::{AppState, CommandEnv};
use crate::TonkWorkerError;
use crate::worker::TonkState;

/// How long after an element asked for a document the sync drain keeps
/// checking its remote cell. Documents nobody has open are not polled.
const REQUEST_WINDOW_SECONDS: f64 = 30.0;

/// Path parameters of the document route.
#[derive(Debug, Deserialize)]
pub struct DocumentPath {
    /// The repository.
    pub repo: String,
    /// The branch whose heads claim selects the version.
    pub branch: String,
    /// The document entity.
    pub entity: String,
}

/// Query parameters of a read.
#[derive(Debug, Default, Deserialize)]
pub struct ReadParams {
    /// The format to create the document with when the entity is not
    /// one yet. An element names its own; without it a missing document
    /// is 404.
    pub format: Option<String>,
}

/// The body of a write.
#[derive(Debug, Deserialize)]
pub struct WriteRequest {
    /// The heads the writer last saw. Absent = the branch's own.
    #[serde(default)]
    pub heads: Option<Vec<String>>,
    /// The edits, applied as one automerge change.
    pub edits: Vec<Edit>,
    /// The format to create with, as in [`ReadParams::format`].
    #[serde(default)]
    pub format: Option<String>,
}

/// A document as one branch sees it, on the wire.
#[derive(Debug, Serialize)]
pub struct DocumentResponse {
    /// `automerge/text@1` or `automerge/table@1`.
    pub format: &'static str,
    /// The heads the branch is at.
    pub heads: Vec<String>,
    /// After a write: the heads the writer's own content corresponds to.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub local: Option<Vec<String>>,
    /// The markdown of a text document.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub text: Option<String>,
    /// The workbook of a table document.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub table: Option<Table>,
}

impl DocumentResponse {
    fn new(snapshot: Snapshot, local: Option<Vec<String>>) -> Self {
        let (text, table) = match snapshot.content {
            Content::Text(text) => (Some(text), None),
            Content::Table(table) => (None, Some(table)),
        };
        Self {
            format: snapshot.format.name(),
            heads: snapshot.heads,
            local,
            text,
            table,
        }
    }

    fn etag(&self) -> String {
        format!("\"{}\"", self.heads.join("."))
    }
}

type Key = (String, String, String);

#[derive(Default)]
struct Tracker {
    /// When an element last asked for a document, unix seconds.
    requested: HashMap<Key, f64>,
    /// Documents written here since their last successful sync pass.
    dirty: HashSet<Key>,
    /// Branches whose documents were scanned once this worker lifetime:
    /// the dirty set does not survive a restart, the cells do.
    scanned: HashSet<(String, String)>,
}

fn tracker() -> &'static Mutex<Tracker> {
    static TRACKER: OnceLock<Mutex<Tracker>> = OnceLock::new();
    TRACKER.get_or_init(Mutex::default)
}

fn now_seconds() -> f64 {
    #[cfg(all(target_arch = "wasm32", target_os = "unknown"))]
    {
        js_sys::Date::now() / 1000.0
    }
    #[cfg(not(all(target_arch = "wasm32", target_os = "unknown")))]
    {
        std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .map_or(0.0, |elapsed| elapsed.as_secs_f64())
    }
}

fn key(repo: &str, branch: &str, entity: &Entity) -> Key {
    (repo.to_string(), branch.to_string(), entity.to_string())
}

fn note_requested(repo: &str, branch: &str, entity: &Entity) {
    if let Ok(mut tracker) = tracker().lock() {
        tracker.requested.insert(key(repo, branch, entity), now_seconds());
    }
}

fn note_dirty(repo: &str, branch: &str, entity: &Entity) {
    if let Ok(mut tracker) = tracker().lock() {
        tracker.dirty.insert(key(repo, branch, entity));
    }
}

fn parse_entity(text: &str) -> Result<Entity, TonkWorkerError> {
    text.parse()
        .map_err(|error| TonkWorkerError::Router(format!("Invalid document entity '{text}': {error}")))
}

fn parse_format(name: Option<&str>) -> Result<Option<Format>, TonkWorkerError> {
    match name {
        None => Ok(None),
        Some(name) => Format::parse(name)
            .map(Some)
            .ok_or_else(|| TonkWorkerError::Router(format!("Unknown document format '{name}'"))),
    }
}

fn session_error(error: session::SessionError) -> TonkWorkerError {
    use session::SessionError;
    match error {
        SessionError::NotADocument(_) => TonkWorkerError::NotFound(error.to_string()),
        SessionError::UnknownFormat(..) => TonkWorkerError::Conflict(error.to_string()),
        SessionError::Document(_) => TonkWorkerError::Router(error.to_string()),
        other => TonkWorkerError::Internal(other.to_string()),
    }
}

/// The writer's stamp: this profile, now.
fn stamp(tonk: &TonkState) -> Stamp {
    Stamp {
        author: Some(tonk.profile.did().to_string()),
        time: now_seconds() as i64,
    }
}

/// Acquire the reactor's session for `repo`/`branch`; an empty `repo`
/// is the profile repository, as in [`super::CommandOrigin`].
async fn acquire(
    tonk: &TonkState,
    repo: &str,
    branch: &str,
) -> Result<dialog_reactor::BranchSession, TonkWorkerError> {
    let reference = if repo.is_empty() {
        tonk.reactor.profile_repository().branch(branch)
    } else {
        tonk.reactor.repository(repo).branch(branch)
    };
    reference
        .acquire(&tonk.operator)
        .await
        .map_err(|error| TonkWorkerError::NotFound(format!("branch {repo}/{branch}: {error}")))
}

/// Refresh the mirror of `entity` and let subscribers know.
async fn refresh_mirror(tonk: &TonkState, session: &dialog_reactor::BranchSession, entity: &Entity) {
    match session::mirror(session.handle(), entity, &tonk.operator).await {
        Ok(_) => {
            tonk.reactor.schedule_poll(Arc::clone(&session.state));
            tonk.reactor.run_scheduled_polls(&tonk.operator).await;
        }
        Err(error) => log!("document mirror of {entity} failed: {error}"),
    }
}

/// `GET /api/repository/{repo}/branch/{branch}/document/{entity}`
///
/// The document as this branch sees it. `If-None-Match` with the last
/// ETag answers 304 when the branch's heads did not move; the request
/// never leaves the device, so an element can watch cheaply.
#[wasm_compat]
pub async fn read(
    State(state): State<AppState>,
    Path(path): Path<DocumentPath>,
    Query(params): Query<ReadParams>,
    headers: HeaderMap,
) -> Result<Response, TonkWorkerError> {
    let entity = parse_entity(&path.entity)?;
    let create = parse_format(params.format.as_deref())?;
    let tonk = state.read().await;
    let session = acquire(&tonk, &path.repo, &path.branch).await?;
    note_requested(&path.repo, &path.branch, &entity);

    let before = session.handle().revision();
    let snapshot = session::read(session.handle(), &entity, create, &tonk.operator)
        .await
        .map_err(session_error)?;
    if session.handle().revision() != before {
        // The open declared or converted the document.
        note_dirty(&path.repo, &path.branch, &entity);
        tonk.sync_queue
            .mark_dirty(&path.repo, super::sync::now_millis());
        refresh_mirror(&tonk, &session, &entity).await;
    }

    let body = DocumentResponse::new(snapshot, None);
    let etag = body.etag();
    let unchanged = headers
        .get(header::IF_NONE_MATCH)
        .and_then(|value| value.to_str().ok())
        .is_some_and(|seen| seen == etag);
    if unchanged {
        return Ok((StatusCode::NOT_MODIFIED, [(header::ETAG, etag)]).into_response());
    }
    Ok(([(header::ETAG, etag)], Json(body)).into_response())
}

/// `POST /api/repository/{repo}/branch/{branch}/document/{entity}`
///
/// Apply edits on top of the heads the writer last saw.
#[wasm_compat]
pub async fn write(
    State(state): State<AppState>,
    Path(path): Path<DocumentPath>,
    body: Bytes,
) -> Result<Response, TonkWorkerError> {
    let entity = parse_entity(&path.entity)?;
    let request: WriteRequest = serde_json::from_slice(&body)
        .map_err(|error| TonkWorkerError::Router(format!("invalid document write: {error}")))?;
    let create = parse_format(request.format.as_deref())?;
    let tonk = state.read().await;
    let session = acquire(&tonk, &path.repo, &path.branch).await?;
    note_requested(&path.repo, &path.branch, &entity);

    let written = session::write(
        session.handle(),
        &entity,
        create,
        request.heads.as_deref(),
        &request.edits,
        &stamp(&tonk),
        &tonk.operator,
    )
    .await
    .map_err(session_error)?;

    note_dirty(&path.repo, &path.branch, &entity);
    tonk.sync_queue
        .mark_dirty(&path.repo, super::sync::now_millis());
    refresh_mirror(&tonk, &session, &entity).await;

    let body = DocumentResponse::new(written.snapshot, Some(written.local));
    let etag = body.etag();
    Ok(([(header::ETAG, etag)], Json(body)).into_response())
}

/// Run one document command on the branch it fired in. The behaviour
/// is `tonk_document::command::run`, the same function the CLI calls;
/// this wrapper adds what only the worker has — the sync queue, the
/// mirror and the subscription poll. A refused edit (no match, two
/// matches, a bad path) changes nothing and is reported as an overlay
/// fact on the command's own entity, so the page that asked can read
/// why.
async fn run(env: &CommandEnv, request: tonk_document::command::Request) {
    let origin = env.origin().clone();
    let document = request.document.clone();
    let tonk = env.state().read().await;
    let session = match acquire(&tonk, &origin.repo, &origin.branch).await {
        Ok(session) => session,
        Err(error) => return log!("document command on {document}: {error}"),
    };
    let result =
        tonk_document::command::run(session.handle(), &tonk.operator, &stamp(&tonk), &request).await;
    match result {
        Ok(_) => {
            note_dirty(&origin.repo, &origin.branch, &document);
            if !origin.repo.is_empty() {
                tonk.sync_queue
                    .mark_dirty(&origin.repo, super::sync::now_millis());
            }
            refresh_mirror(&tonk, &session, &document).await;
        }
        Err(error) => {
            log!("document command on {document} refused: {error}");
            session.state.assert_overlay(session::failure(
                &request.command,
                &document,
                &error.to_string(),
            ));
            tonk.reactor.schedule_poll(Arc::clone(&session.state));
            tonk.reactor.run_scheduled_polls(&tonk.operator).await;
        }
    }
}

macro_rules! document_provider {
    ($command:ty) => {
        #[cfg_attr(not(target_arch = "wasm32"), async_trait::async_trait)]
        #[cfg_attr(target_arch = "wasm32", async_trait::async_trait(?Send))]
        impl dialog_capability::Provider<$command> for CommandEnv {
            async fn execute(&self, command: $command) {
                run(self, command.into()).await;
            }
        }
    };
}

document_provider!(tonk_schema::command::DocumentReplace);
document_provider!(tonk_schema::command::DocumentInsert);
document_provider!(tonk_schema::command::DocumentSplice);
document_provider!(tonk_schema::command::DocumentPut);
document_provider!(tonk_schema::command::DocumentRemove);
document_provider!(tonk_schema::command::DocumentRestore);

/// The document half of a repository's sync sweep: run the sync pass
/// for every document of `branch` that is dirty or that an element
/// asked for lately. Called after the branch itself reconciled, so the
/// heads claims a pull brought in can be shown as soon as their bytes
/// land.
pub(crate) async fn sync_documents(state: &AppState, repo: &str, branch: &str) -> Result<(), String> {
    let tonk = state.read().await;
    let session = acquire(&tonk, repo, branch)
        .await
        .map_err(|error| error.to_string())?;
    let handle = session.handle();

    let first_scan = tracker()
        .lock()
        .map(|mut tracker| tracker.scanned.insert((repo.to_string(), branch.to_string())))
        .unwrap_or(false);

    let mut candidates: HashSet<String> = HashSet::new();
    if let Ok(tracker) = tracker().lock() {
        let now = now_seconds();
        for ((r, b, entity), at) in &tracker.requested {
            if r == repo && b == branch && now - at <= REQUEST_WINDOW_SECONDS {
                candidates.insert(entity.clone());
            }
        }
        for (r, b, entity) in &tracker.dirty {
            if r == repo && b == branch {
                candidates.insert(entity.clone());
            }
        }
    }
    if first_scan {
        // Claims-era prose bodies and workbooks convert once: a
        // document-mode view matches on the format claim, so until then
        // an old entity renders nothing.
        match session::adopt_legacy(handle, &tonk.operator).await {
            Ok(0) => {}
            Ok(converted) => {
                log!("converted {converted} claims-era documents in {repo}/{branch}");
                tonk.sync_queue.mark_dirty(repo, super::sync::now_millis());
            }
            Err(error) => log!("document conversion in {repo}/{branch} failed: {error}"),
        }
        // After a restart the dirty set is gone but the cells are not:
        // find unpublished work once, and fill the mirror while here.
        match session::documents(handle, &tonk.operator).await {
            Ok(documents) => {
                for (entity, _) in documents {
                    if session::is_dirty(handle, &entity, &tonk.operator)
                        .await
                        .unwrap_or(false)
                    {
                        candidates.insert(entity.to_string());
                    }
                    refresh_mirror(&tonk, &session, &entity).await;
                }
            }
            Err(error) => log!("document scan of {repo}/{branch} failed: {error}"),
        }
    }

    let mut failed = false;
    for text in candidates {
        let Ok(entity) = text.parse::<Entity>() else {
            continue;
        };
        match session::sync(handle, &entity, &tonk.operator).await {
            Ok(outcome) => {
                if let Ok(mut tracker) = tracker().lock() {
                    tracker.dirty.remove(&key(repo, branch, &entity));
                }
                if outcome.pulled {
                    refresh_mirror(&tonk, &session, &entity).await;
                }
            }
            Err(error) => {
                log!("document sync of {entity} in {repo}/{branch} failed: {error}");
                failed = true;
            }
        }
    }
    if failed {
        Err(format!("documents of '{repo}/{branch}' did not fully sync"))
    } else {
        Ok(())
    }
}

#[cfg(all(test, not(all(target_arch = "wasm32", target_os = "unknown"))))]
mod tests {
    use super::*;
    use crate::router::command::tests::native::test_state;
    use crate::router::{CommandOrigin, dispatch};
    use ::axum::body::to_bytes;
    use dialog_artifacts::{Changes, Statement as _};
    use dialog_query::{Output as _, Query as ConceptQuery, Term, the};
    use tonk_schema::Replica;
    use tonk_schema::prelude::DidExt as _;

    /// Create a space through the real command and return its repo key.
    async fn space(state: &AppState) -> String {
        let mut changes = Changes::new();
        the!("xyz.tonk.command.create-space/name")
            .of("cmd:create".parse::<Entity>().unwrap())
            .is("Documents".to_string())
            .assert(&mut changes);
        dispatch(state, CommandOrigin::default(), changes).await;

        let tonk = state.read().await;
        let meta = tonk
            .reactor
            .profile_repository()
            .branch("main")
            .acquire(&tonk.operator)
            .await
            .unwrap();
        let rows: Vec<Replica> = meta
            .handle()
            .query()
            .select(ConceptQuery::<Replica> {
                this: Term::var("this"),
                subject: Term::var("subject"),
                profile: Term::var("profile"),
                kind: Term::var("kind"),
            })
            .perform(&tonk.operator)
            .try_vec()
            .await
            .unwrap();
        let subject: dialog_varsig::Did = rows
            .into_iter()
            .find(|replica| replica.kind == Replica::repository_kind())
            .and_then(|replica| replica.subject.0.to_string().parse().ok())
            .expect("the space was created");
        subject.repo_key().to_owned()
    }

    fn path(repo: &str) -> Path<DocumentPath> {
        Path(DocumentPath {
            repo: repo.to_string(),
            branch: "main".to_string(),
            entity: "id:prose/doc".to_string(),
        })
    }

    async fn json(response: Response) -> serde_json::Value {
        let bytes = to_bytes(response.into_body(), usize::MAX).await.unwrap();
        serde_json::from_slice(&bytes).unwrap()
    }

    async fn post(state: &AppState, repo: &str, body: serde_json::Value) -> serde_json::Value {
        let response = write(State(state.clone()), path(repo), Bytes::from(body.to_string()))
            .await
            .expect("the write is accepted");
        json(response).await
    }

    fn heads(value: &serde_json::Value, field: &str) -> Vec<String> {
        serde_json::from_value(value[field].clone()).unwrap()
    }

    #[dialog_common::test]
    async fn it_serves_a_document_to_an_element_and_runs_a_command_on_it() {
        let state = test_state().await;
        let repo = space(&state).await;

        // A missing document is 404 until a caller names a format.
        let missing = read(State(state.clone()), path(&repo), Query(ReadParams::default()), HeaderMap::new()).await;
        assert!(matches!(missing, Err(TonkWorkerError::NotFound(_))));

        // The element's first write creates it.
        let first = post(
            &state,
            &repo,
            serde_json::json!({ "format": "automerge/text@1", "edits": [{ "edit": "set-text", "text": "hello world" }] }),
        )
        .await;
        assert_eq!(first["text"], "hello world");
        assert_eq!(first["format"], "automerge/text@1");

        // A read returns the same heads; the ETag answers 304 after.
        let response = read(State(state.clone()), path(&repo), Query(ReadParams::default()), HeaderMap::new())
            .await
            .unwrap();
        let etag = response.headers().get(header::ETAG).unwrap().clone();
        let body = json(response).await;
        assert_eq!(heads(&body, "heads"), heads(&first, "heads"));
        let mut seen = HeaderMap::new();
        seen.insert(header::IF_NONE_MATCH, etag);
        let again = read(State(state.clone()), path(&repo), Query(ReadParams::default()), seen)
            .await
            .unwrap();
        assert_eq!(again.status(), StatusCode::NOT_MODIFIED);

        // An agent asserts `document/replace`: the same command a page
        // sends through /transact, dispatched on the space's own branch.
        let mut command = Changes::new();
        let this: Entity = "cmd:replace".parse().unwrap();
        the!("xyz.tonk.document.replace/document")
            .of(this.clone())
            .is("id:prose/doc".parse::<Entity>().unwrap())
            .assert(&mut command);
        the!("xyz.tonk.document.replace/find")
            .of(this.clone())
            .is("world".to_string())
            .assert(&mut command);
        the!("xyz.tonk.document.replace/with")
            .of(this)
            .is("there".to_string())
            .assert(&mut command);
        let origin = CommandOrigin {
            repo: repo.clone(),
            branch: "main".to_string(),
            client: None,
        };
        dispatch(&state, origin, command).await;

        let after = json(
            read(State(state.clone()), path(&repo), Query(ReadParams::default()), HeaderMap::new())
                .await
                .unwrap(),
        )
        .await;
        assert_eq!(after["text"], "hello there", "the command edited the document");

        // The element, which still holds the OLD heads, sends its own
        // edit: both survive.
        let merged = post(
            &state,
            &repo,
            serde_json::json!({ "heads": heads(&first, "heads"), "edits": [{ "edit": "set-text", "text": "oh hello world" }] }),
        )
        .await;
        assert_eq!(merged["text"], "oh hello there");
        assert_ne!(heads(&merged, "local"), heads(&merged, "heads"));
    }
}
